// waku/dist/adapters/node.js
import path from 'node:path';
import { serve } from '@hono/node-server';
import { serveStatic } from '@hono/node-server/serve-static';
import { Hono } from 'hono';
import { unstable_constants as constants, unstable_createServerEntryAdapter as createServerEntryAdapter, unstable_honoMiddleware as honoMiddleware } from 'waku/internals';
const { DIST_PUBLIC } = constants;
const { contextMiddleware, rscMiddleware, middlewareRunner } = honoMiddleware;
export default createServerEntryAdapter(({ processRequest, processBuild, config, isBuild, notFoundHtml }, options)=>{
    const { middlewareFns = [], middlewareModules = {} } = options || {};
    const app = new Hono();
    app.notFound((c)=>{
        if (notFoundHtml) {
            return c.html(notFoundHtml, 404);
        }
        return c.text('404 Not Found', 404);
    });
    if (isBuild) {
        app.use(`${config.basePath}*`, serveStatic({
            root: path.join(config.distDir, DIST_PUBLIC),
            rewriteRequestPath: (path)=>path.slice(config.basePath.length - 1)
        }));
    }
    app.use(contextMiddleware());
    for (const middlewareFn of middlewareFns){
        app.use(middlewareFn());
    }
    app.use(middlewareRunner(middlewareModules));
    app.use(rscMiddleware({
        processRequest
    }));
    const buildOptions = {
        distDir: config.distDir
    };
    return {
        fetch: app.fetch,
        build: processBuild,
        buildOptions,
        buildEnhancers: [
            'waku/adapters/node-build-enhancer'
        ],
        serve
    };
});
