// waku/dist/adapters/vercel.js
import { getRequestListener } from '@hono/node-server';
import { Hono } from 'hono';
import { unstable_constants as constants, unstable_createServerEntryAdapter as createServerEntryAdapter, unstable_honoMiddleware as honoMiddleware } from 'waku/internals';
const { DIST_PUBLIC } = constants;
const { contextMiddleware, rscMiddleware, middlewareRunner } = honoMiddleware;
global.__WAKU_HONO_NODE_SERVER_GET_REQUEST_LISTENER__ = getRequestListener;
export default createServerEntryAdapter(({ processRequest, processBuild, config, notFoundHtml }, options)=>{
    const { middlewareFns = [], middlewareModules = {} } = options || {};
    const app = new Hono();
    app.notFound((c)=>{
        if (notFoundHtml) {
            return c.html(notFoundHtml, 404);
        }
        return c.text('404 Not Found', 404);
    });
    app.use(contextMiddleware());
    for (const middlewareFn of middlewareFns){
        app.use(middlewareFn());
    }
    app.use(middlewareRunner(middlewareModules));
    app.use(rscMiddleware({
        processRequest
    }));
    const buildOptions = {
        assetsDir: options?.assetsDir || 'assets',
        distDir: config.distDir,
        rscBase: config.rscBase,
        privateDir: config.privateDir,
        basePath: config.basePath,
        DIST_PUBLIC,
        serverless: !options?.static
    };
    return {
        fetch: app.fetch,
        build: processBuild,
        buildOptions,
        buildEnhancers: [
            'waku/adapters/vercel-build-enhancer'
        ]
    };
});
